### ISSUE TITLE
Set anti-affinity to replicated/HA pods

### ISSUE THREAD

**Issue by fabxc** [coreos/prometheus-operator#127]
We should do this automatically. How does it interact with user-defined nodeSelectors or, if later added, user-defined anti-affinity? Does the latter just overwrite it and the user has to ensure its sanity?

**Comment by fabxc** [comment]
In my benchmark cluster I use the following annotation which seems to work. (Can always work just by accident of course.)

```yaml
scheduler.alpha.kubernetes.io/affinity: >
  {
    "podAntiAffinity": {
      "requiredDuringSchedulingIgnoredDuringExecution": [
        {
          "labelSelector": {
            "matchExpressions": [
              {
                "key": "app",
                "operator": "In",
                "values": [
                  "prometheus"
                ]
              }
            ]
          },
          "topologyKey": "kubernetes.io/hostname"
        }
      ]
    }
  }
```

Label matcher has to also cover the specific `prometheus` label of course.

**Comment by galexrt** [comment]
If this should be added it, having an option to toggle it would be good so the user can choose when he wants it. Also I think the default should be `true` for this option.

**Comment by brancz** [comment]
It should be free style choose-able by the user entirely as we can't make assumptions about node labels of a users cluster nodes, however, being an alpha feature I would prefer waiting for `1.6` where this feature is promoted to beta.

**Comment by fabxc** [comment]
Mh, yea, maybe we just want to make part of the TPR. It's one of those things though that I'd just enable for everyone initially and then think about extending our complexity if a use case comes up.

HA is only really done to prevent against node failure. So I can't think of any good reason why one would have two Prometheis scheduled on the same node.

In that case node labels are orthogonal as the `kubernetes.io/hostname` one is default afaik.

But yea, people might have additional node labels or might want to go as far as scheduling them on a node in a different zone. Let's get back to this once its integrated into the main YAML spec.

**Comment by brancz** [comment]
We'll be able to set the anti-affinity directly on the pod now, with 1.6. We just have to wait for client-go to be released with 1.6 compatibility.

**Comment by gianrubio** [comment]
That the same behaviour of [etcd-operator afinity](https://github.com/coreos/etcd-operator/search?utf8=✓&q=affinity&type=), it's just a boolean flag to choose to use affinity or not. I really like this idea, it's simple for the users also solve the HA problem. 

The latest release of client-go `podSpec`  support [affinity](https://github.com/kubernetes/client-go/blob/master/pkg/api/v1/types.go#L2401).

**Comment by brancz** [comment]
I was hoping there would be a non beta release of client-go some time soon, but it looks like it is taking a while, so we should probably update to 2.0-beta.

**Comment by brancz** [comment]
as we upgraded client-go this is now possible, @fabxc what do you think of @gianrubio's idea of not exposing it directly but rather doing the "HA default"?

**Comment by gianrubio** [comment]
I imagine this as a boolean flag `antiAffinity` and a string flag `antiAffinityTopologyKey` so the user are able to choose the `topologyKey`.

**Comment by brancz** [comment]
We're currently playing with the idea of using `PodPreset`s for any configuration that is not Prometheus specific, eg. any scheduling configuration.

**Comment by brancz** [comment]
As initializers and presets are still in alpha, we decided to move forward with this and implemented it into the object directly afterall. (closed by #600)
